[SPARK-10101] [SQL]  Add maxlength to JDBC field metadata and override JDBCDialects for strings as VARCHAR.

rama-mullapudi opened the pull request.

rama-mullapudi committed:
Added stringDataType option to jdbc connection properties

AmplabJenkins commented:
Can one of the admins verify this patch?

rxin commented:
I think the current solution is to use the jdbc dialect, similar to the pull request here: https://github.com/apache/spark/pull/8393/files

Are you seeing a database that we don't support yet? Maybe we can just add a dialect for it.

rama-mullapudi commented:
I was using Oracle and IBM Netezza databases to write to, I thought of adding Jdbc dialect but in case of strings CLOB is the one without character limit which can be added to jdbc dialect but most databases i.e Oracle, DB2, Teradata have restrictions on usage of CLOB columns ex can not use CLOB columns in group by or distinct. Since VARCHAR(n) is the most preferred option for these databases and in most cases its VARCHAR(255) or less, what would be best way to implement string to VARCHAR conversion.
My idea is to provide an optional jdbc connection parameter so developer can decide to use VARCHAR(20) or VARCHAR(2000) or CLOB and can pass it as jdbc connection options and all string columns for that table will use the data type passed by developer.

marmbrus commented:
In spark-redshift we are planning to use an optional metadata field on the column `maxLength` that when present will let the data source use a fixed length column.  I would be good to standardize here.

rama-mullapudi commented:
Since String to varchar(n) is common problem across multiple databases, would it be possible to push spark-redshift change to spark core so every one can benifit from it. Should I go ahead and close this pull request so we can have more flexible solution.

marmbrus commented:
I don't think there is anything in conflict with implementing this in both libraries.  Really the redshift library is just a simplified version of this since it only needs to handle one dialect (the reason it is its own library is because it actually does extra work using S3 to parallel extract data instead of using JDBC as the channel).

I'm only suggestion that we use the same option so that users don't have to learn two different concepts.  Can we just add some logic like this (https://github.com/databricks/spark-redshift/pull/54/files#diff-69806564231efb590460b162532ba683R145) in the appropriate dialects here in spark core?

rama-mullapudi committed:
Added maxlength for field metadata so string types can use for VARCHAR.
Changed  getJDBCType to take 2 parameters DataType and MetaData

Usage in Scala
import org.apache.spark.sql.types.MetadataBuilder
val metadata = new MetadataBuilder().putLong("maxlength", 10).build()
df.withColumn("colName", col("colName").as("colName", metadata)

rama-mullapudi committed:
Added maxlength for field metadata so string types can use for VARCHAR.
Changed  getJDBCType to take 2 parameters DataType and MetaData

Usage in Scala
import org.apache.spark.sql.types.MetadataBuilder
val metadata = new MetadataBuilder().putLong("maxlength", 10).build()
df.withColumn("colName", col("colName").as("colName", metadata)

rama-mullapudi committed:
Added maxlength for field metadata so string types can use for VARCHAR.
Changed  getJDBCType to take 2 parameters DataType and MetaData

Usage in Scala
import org.apache.spark.sql.types.MetadataBuilder
val metadata = new MetadataBuilder().putLong("maxlength", 10).build()
df.withColumn("colName", col("colName").as("colName", metadata)

rama-mullapudi commented:
Added maxlength for field metadata so string types can be converted to VARCHAR(n).

Added JDBCDialects for OracleDialect, NetezzaDialect and modifed DB2Dialect to use VARCHAR when maxlength exists.

Modified existing getJDBCType function to take 2 parameters DataType and MetaData, so maxlength metadata can be used in JDBCDialects.

Can you review the change.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:
```diff
@@ -77,9 +77,10 @@ abstract class JdbcDialect {
    * @return The new JdbcType if there is an override for this DataType
    */
-  def getJDBCType(dt: DataType): Option[JdbcType] = None
+  def getJDBCType(dt: DataType, md: Metadata): Option[JdbcType] = None
```
While this is a `DeveloperAPI`, it is public so it would be good to fix without breaking binary compatibility.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:
```diff
@@ -161,13 +161,15 @@ object JdbcUtils extends Logging {
-  def schemaString(df: DataFrame, url: String): String = {
+  def schemaString(
+                    df: DataFrame,
+                    url: String): String = {
```
This doesn't need to change (and is not indented incorrectly).
https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-Indentation

pallavipr commented:
DB2 team is working on its own dialect that works with different flavors of DB2. Code will shortly be submitted.

rama-mullapudi committed:
Added method override for getJDBCType to take 2 parameters DataType and MetaData and removed DB2 JdbcDialect as DB2 team is working on the dialect.

rama-mullapudi commented:
Left getJDBCType with single parameter for binary compatibility and removed DB2 dialect for DB2 team. Can you take a look.

rama-mullapudi committed:
Sync to master

Conflicts:
	sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala
	sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala

pallavipr commented:
Looks good Rama. We are almost done with DB2 changes - will send for review soon.

One question, did you introduce a stringDataType property in connection url? And StringType will be mapped to the value provided for stringDataType?

Thanks,
Pallavi

rama-mullapudi committed:
Sync Master

rama-mullapudi committed:
Sync Master

rama-mullapudi committed:
Updated a call with added parameter

rama-mullapudi commented:
Can one of admins review changes

marmbrus commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43054 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43054/consoleFull) for   PR 8374 at commit [`dddc137`](https://github.com/apache/spark/commit/dddc13762fc97a2dd8e407729db19cfee4cf6ccb).

SparkQA commented:
[Test build #43054 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43054/console) for   PR 8374 at commit [`dddc137`](https://github.com/apache/spark/commit/dddc13762fc97a2dd8e407729db19cfee4cf6ccb).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43054/
Test FAILed.

sureshthalamati commented:
I am new to git,  not sure if I am  reading the changes correctly.  It looks you 
are undoing https://issues.apache.org/jira/browse/SPARK-9078 changes in your last 
merge. 

I see you fixed the current dialect implementation in the code to use the new method, but there 
can be custom dialect implementations using the old  getJDBCType(dt: DataType) method.  
After this change , user specified existing custom dialects may not work. 

Example :
import org.apache.spark.sql.jdbc._
JdbcDialects.registerDialect(MyNetEzzaDialect)

Another minor thing, any particualar reason for setting java.sql.Types.CHAR insead of 
java.sql.Types.VARCHAR in the following statements. 

OracleDialect :
 Some(JdbcType(s"VARCHAR(${md.getLong("maxlength")})", java.sql.Types.CHAR))
NetEzzaDialect :
 Some(JdbcType("VARCHAR(255)", java.sql.Types.CHAR))

rama-mullapudi commented:
Intent of the change is make field metadata available to getJDBCType so it can be used by dialects to handle special cases. One I was trying to solve is by default java string types are stored as TEXT data type in database and most databases doesn't support TEXT data type, so trying to use VARCHAR which needs max length to create a varchar field in database.

What would be the best way to handle custom dialects, since its developerapi and can change.

Thanks for looking in to the change and will update code to use java.sql.Types.VARCHAR instead of java.sql.Types.CHAR

rama-mullapudi committed:
Fixed spark code style where comment exceeded 100 chars, and changed java.sql.Types.CHAR to java.sql.Types.VARCHAR when type used is VARCHAR

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43063 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43063/consoleFull) for   PR 8374 at commit [`5f532e8`](https://github.com/apache/spark/commit/5f532e8c0de5fab2f354d15be414d642ffc35aef).

SparkQA commented:
[Test build #43063 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43063/console) for   PR 8374 at commit [`5f532e8`](https://github.com/apache/spark/commit/5f532e8c0de5fab2f354d15be414d642ffc35aef).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43063/
Test FAILed.

pallavipr commented:
Rama, while specifying DataFrame metadata length is good approach, can we please have dbStringMappingType and dbStringMappingLength as properties on the connection url as well? It will make it really easy for developers, as opposed to dealing with dataframe length.

We have it implemented for IBM data servers, trying to see if we can make it mainline Spark since its a useful feature.

VARCHAR has its limitations on length, hence under many circumstances, users may want to map StringType to CLOB of varying lengths if data server supports.

Let me know if we can pursue - we have the code ready.

Thanks.
Pallavi

rama-mullapudi committed:
Merge remote-tracking branch 'upstream/master'

marmbrus commented:
Just to let you know, we are busy wrapping up 1.5.1, but I have put reviewing this PR on our schedule for the next 2 week sprint.

rama-mullapudi committed:
Fixed sync with master issue

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43068 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43068/consoleFull) for   PR 8374 at commit [`27f118b`](https://github.com/apache/spark/commit/27f118be2b6c4bf1ce34b40d79691f17856e0bba).

SparkQA commented:
[Test build #43068 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43068/console) for   PR 8374 at commit [`27f118b`](https://github.com/apache/spark/commit/27f118be2b6c4bf1ce34b40d79691f17856e0bba).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43068/
Test FAILed.

rama-mullapudi committed:
Fixed JDBCSuite for DB2 to send extra parameter for getJDBCType

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43071 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43071/consoleFull) for   PR 8374 at commit [`44e1978`](https://github.com/apache/spark/commit/44e1978530a9703ecad9fae90a7ea9ebb0693fd3).

SparkQA commented:
[Test build #43071 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43071/console) for   PR 8374 at commit [`44e1978`](https://github.com/apache/spark/commit/44e1978530a9703ecad9fae90a7ea9ebb0693fd3).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43071/
Test FAILed.

rama-mullapudi committed:
Fixed JDBCSuite for DB2 to send extra parameter for getJDBCType

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43074 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43074/consoleFull) for   PR 8374 at commit [`e605a11`](https://github.com/apache/spark/commit/e605a11383f2a0c22003e3a3cdf349b26af9310f).

SparkQA commented:
**[Test build #43074 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43074/console)**     for PR 8374 at commit [`e605a11`](https://github.com/apache/spark/commit/e605a11383f2a0c22003e3a3cdf349b26af9310f)     after a configured wait of `250m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43074/
Test FAILed.

rama-mullapudi committed:
Fixed JDBCSuite for DB2 to send extra parameter for getJDBCType

rama-mullapudi committed:
Updated JDBCDialects to save VARCHAR size to metadata

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43089 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43089/consoleFull) for   PR 8374 at commit [`4c2a7a4`](https://github.com/apache/spark/commit/4c2a7a43754485ad5771250ebaf8991a0137df0b).

SparkQA commented:
[Test build #43089 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43089/console) for   PR 8374 at commit [`4c2a7a4`](https://github.com/apache/spark/commit/4c2a7a43754485ad5771250ebaf8991a0137df0b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43089/
Test FAILed.

rama-mullapudi committed:
Updated JDBCDialects to save VARCHAR size to metadata

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43099 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43099/consoleFull) for   PR 8374 at commit [`a0cb024`](https://github.com/apache/spark/commit/a0cb024b74c602e531f51020cb1a1ee97f8238ff).

sureshthalamati commented:
The new contract defined in the JdbcDialect class  is sub classes can define any one of the getJDBCType methods or both.  Both the methods has to be called  to check if dialect has specified different mapping.   I think it makes sense to put the call with metadata first, and if it is not defined then call the one without the metadata.  If we address this, any existing custom dialects I mentioned in my previous comment should also work fine.
 
I think Code change in JdbcUtils.scala  that call getJDBCType() has to be changed to something like the following:

val typ: String = dialect.getJDBCType(field.dataType, field.metadata).map(_.databaseTypeDefinition).orElse(dialect.getJDBCType(field.dataType).map(_.databaseTypeDefinition)).getOrElse(
field.dataType match { …

I hope that helps . Thank you for working on this issue.

SparkQA commented:
[Test build #43099 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43099/console) for   PR 8374 at commit [`a0cb024`](https://github.com/apache/spark/commit/a0cb024b74c602e531f51020cb1a1ee97f8238ff).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43099/
Test PASSed.

rama-mullapudi committed:
Changed call for getJDBCType to call both methods with datatype and metadata and one with only datatype.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43224 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43224/consoleFull) for   PR 8374 at commit [`d50bdf7`](https://github.com/apache/spark/commit/d50bdf796c96f827aa920c31dd3d5822414f178d).

SparkQA commented:
[Test build #43224 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43224/console) for   PR 8374 at commit [`d50bdf7`](https://github.com/apache/spark/commit/d50bdf796c96f827aa920c31dd3d5822414f178d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43224/
Test PASSed.

rama-mullapudi commented:
Please review the pull for approval to merge.

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:
```diff
@@ -82,6 +82,14 @@ abstract class JdbcDialect {
+   * @param md The metadata
+   * @return The new JdbcType if there is an override for this DataType
+   */
+  def getJDBCType(dt: DataType, md: Metadata): Option[JdbcType] = None
```
How about calling `getJDBCType(dt,Metadata.empty)` instead of `None` in `getJDBCType(dt: DataType)`, because the classes that extends `JdbcDialect` possibly implement one of them and then the behaviours of the two function totally different?

rama-mullapudi replied:
Sorry did not get the change you are suggesting, do you mean to call getJDBCType(dt,Metadata.empty) from getJDBCType(dt: DataType).

maropu replied:
@rama-mullapudi yes.

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:
```diff
@@ -278,3 +285,59 @@ case object MsSqlServerDialect extends JdbcDialect {
+  override def getJDBCType(dt: DataType, md: Metadata): Option[JdbcType] = {
+    if (dt == StringType && md.contains("maxlength")) {
+      Some(JdbcType(s"VARCHAR(${md.getLong("maxlength")})", java.sql.Types.VARCHAR))
+    } else if (dt == StringType ) {
```
Nit: Remove the end single space in parenthesis.

maropu commented:
Great work! I left some trivial comments.

pallavipr commented:
Rama, can we please get the metadata maxlength changes in? We have a serious problem with writing Stringtypes, and being able to map according to Dataframe maxlength would be an alternative to get around this issue.

Thanks.

rishitesh commented:
Is this PR going to be merged shortly. We are also expecting this PR in so that we don't have to change Spark. Can anybody respond whats the status ? If any help needed do let me know.

maropu commented:
Seems no because it gets stale. If nobody takes this, I'll do it.

HyukjinKwon commented:
It seems this PR is stale..

HyukjinKwon commented:
ping @rama-mullapudi

robbyki commented:
Apologies for misunderstanding this issue but I'm going through several resources to try and understand how to maintain my schema created outside of spark and then just truncating my tables from spark followed by writing with a savemode of overwrite. My problem exactly this issue with respect to my db netezza failing when it sees spark trying to save a text data type so I then have to go specify in my new jdbc dialect to use varchar(n) which does work however that just replaces all of my varchar columns (different lengths for different columns) with whatever I specified in my dialect which is not what I want. How can I just have it save the TEXT as varchar without specifying a length in the custom dialect?
